listenbrainz/webserver/static/js/src/stats/UserEntity.tsx

```diff
@@ -0,0 +1,452 @@
+        data = data as UserArtistsResponse;
+        maxListens = data.payload.artists[0].listen_count;
+        totalPages = Math.ceil(
+          data.payload.total_artist_count / this.ROWS_PER_PAGE
+        );
+        entityCount = data.payload.total_artist_count;
+      } else if (newEntity === "release") {
+        data = data as UserReleasesResponse;
+        maxListens = data.payload.releases[0].listen_count;
+        totalPages = Math.ceil(
+          data.payload.total_release_count / this.ROWS_PER_PAGE
+        );
+        entityCount = data.payload.total_release_count;
+      }
+
+      data = await this.getData(page, range, newEntity);
+      this.setState({
+        data: this.processData(data, page, newEntity),
+        entity: newEntity,
+        currPage: page,
+        startDate: new Date(data.payload.from_ts * 1000),
+        totalPages,
+        maxListens,
+        entityCount,
+      });
+    } catch (error) {
+      this.handleError(error);
+    }
+  };
+
+  getData = async (
+    page: number,
+    range: UserEntityAPIRange,
+    entity: Entity
+  ): Promise<UserArtistsResponse | UserReleasesResponse> => {
+    const { user } = this.props;
+    const offset = (page - 1) * this.ROWS_PER_PAGE;
+
+    const data = await this.APIService.getUserEntity(
+      user.name,
+      entity,
+      range,
+      offset,
+      this.ROWS_PER_PAGE
+    );
+    return data;
+  };
+
+  processData = (
+    data: UserArtistsResponse | UserReleasesResponse,
+    page: number,
+    entity?: Entity
+  ): UserEntityData => {
+    if (!entity) {
+      // eslint-disable-next-line no-param-reassign
+      ({ entity } = this.state);
+    }
+    const offset = (page - 1) * this.ROWS_PER_PAGE;
+
+    let result = {} as UserEntityData;
+    if (entity === "artist") {
+      result = (data as UserArtistsResponse).payload.artists
+        .map((elem, idx: number) => {
+          return {
+            id: `${offset + idx + 1}. ${elem.artist_name}`,
+            count: elem.listen_count,
+          };
+        })
+        .reverse();
+    } else if (entity === "release") {
+      result = (data as UserReleasesResponse).payload.releases
+        .map((elem, idx: number) => {
+          return {
+            id: `${offset + idx + 1}. ${elem.release_name}`,
+            count: elem.listen_count,
+          };
+        })
+        .reverse();
+    }
+
+    return result;
+  };
+
+  handleError = (error: Error): void => {
+    // Error Boundaries don't catch errors in async code.
+    // Throwing an error in setState fixes this.
+    // This is a hacky solution but should be fixed with upcoming concurrent mode in React.
+    this.setState(() => {
+      throw error;
+    });
+  };
+
+  render() {
+    const {
+      data,
+      range,
+      entity,
+      entityCount,
+      currPage,
+      maxListens,
+      totalPages,
+      startDate,
+    } = this.state;
+    const prevPage = currPage - 1;
+    const nextPage = currPage + 1;
+    const month = [
```

Comment:
This works. I can suggest a more state-of-the-art solution with the internationalization API that could help in the future if we implement other languages:

`startDate.toLocaleString('default', { month: 'long' });` will give you the month name in the user's default language.
`startDate.toLocaleString('en-us', { month: 'long' });` will do the same but in english (what you want here since the rest of the page is in english).


You can also get the year in the same string if you want: `date.toLocaleString('en-us', { month:'long', year: 'numeric' });` will output `May 2020`

Reply:
Didn't know about this, much more elegant than the previous solution :)